x/mongo/driver/topology/server.go

```diff
@@ -294,27 +285,61 @@ func (s *Server) RequestImmediateCheck() {
 	}
 }
 
+// ProcessHandshakeError handles connection errors during the handshake.
+func (s *Server) ProcessHandshakeError(err error, conn driver.Connection) {
+	//ignore stale error
+	if conn.Stale() {
+		return
+	}
+
+	wrappedConnErr := unwrapConnectionError(err)
+	if wrappedConnErr == nil {
+		return
+	}
+
+	// Since the only kind of ConnectionError we receive from pool.Get will be an initialization
+	// error, we should set the description.Server appropriately.
+	s.updateDescription(description.NewServerFromError(s.address, wrappedConnErr, s.Description().TopologyVersion))
+	s.pool.clear()
+}
+
 // ProcessError handles SDAM error handling and implements driver.ErrorProcessor.
-func (s *Server) ProcessError(err error) {
-	desc := s.Description()
+func (s *Server) ProcessError(err error, conn driver.Connection) {
```

Comment:
This needs a lock. Right now, two connections could concurrently get a state change error, report that they're not stale, and cause the pool to be cleared. The streamable isMaster integration tests don't pass without the lock, so I've already added it to my patch, but I'd prefer to have it in your changes instead because your changes are to reduce race conditions in error handling.

Reply:
Also, would it make sense to do an optimistic check for `err == nil` and return early? Seems weird to go through the whole flow if no error happened.